cranelift/codegen/src/ir/instructions.rs

```diff
@@ -284,6 +285,39 @@ impl InstructionData {
         }
     }
 
+    /// Return the value of an immediate if the instruction has one or `None` otherwise. Only
+    /// immediate values are considered, not global values, constant handles, condition codes, etc.
+    pub fn imm_value(&self) -> Option<DataValue> {
+        match self {
+            &InstructionData::UnaryBool { imm, .. } => Some(DataValue::from(imm)),
+            // 8-bit.
+            &InstructionData::BinaryImm8 { imm, .. }
+            | &InstructionData::BranchTableEntry { imm, .. } => Some(DataValue::from(imm as i8)), // Note the switch from unsigned to signed.
+            // 32-bit
+            &InstructionData::UnaryIeee32 { imm, .. } => Some(DataValue::from(imm)),
+            &InstructionData::HeapAddr { imm, .. } => {
+                Some(DataValue::from(Into::<u32>::into(imm) as i32)) // Note the switch from unsigned to signed.
+            }
+            &InstructionData::Load { offset, .. }
+            | &InstructionData::LoadComplex { offset, .. }
+            | &InstructionData::Store { offset, .. }
+            | &InstructionData::StoreComplex { offset, .. }
+            | &InstructionData::StackLoad { offset, .. }
+            | &InstructionData::StackStore { offset, .. }
+            | &InstructionData::TableAddr { offset, .. } => Some(DataValue::from(offset)),
+            // 64-bit.
+            &InstructionData::UnaryImm { imm, .. }
+            | &InstructionData::BinaryImm64 { imm, .. }
+            | &InstructionData::IntCompareImm { imm, .. } => Some(DataValue::from(imm.bits())),
+            &InstructionData::UnaryIeee64 { imm, .. } => Some(DataValue::from(imm)),
+            // 128-bit.
+            &InstructionData::Shuffle { mask: _, .. } => {
+                unimplemented!("unable to retrieve the 128-bit immediate without the DFG")
```

Comment:
I'm not sure about `unimplemented!()` here -- usually that implies it would be possible but we just haven't done it -- perhaps `None` here, since we wrap `imm_value()` at a higher level? Or better, would it make sense to require a borrow of `dfg.immediates` as an argument to this function? Then we wouldn't need the two-level split logic between this match and `LowerCtx`'s...

Reply:
Yeah, I was dubious about which way to go here. I felt like it was nice that this function didn't depend on the `DataFlowGraph` so I went with `None`.

Reply:
...And documented it some more.